lib/winapi/filemanip.c

```diff
@@ -241,6 +242,130 @@ BOOL MoveFileA (LPCTSTR lpExistingFileName, LPCTSTR lpNewFileName)
+    OBJECT_ATTRIBUTES objectAttributes;
+    IO_STATUS_BLOCK ioStatusBlock;
+    FILE_NETWORK_OPEN_INFORMATION networkOpenInformation;
+    LPVOID readBuffer;
+    DWORD bytesRead;
+
+    RtlInitAnsiString(&sourcePath, lpExistingFileName);
+    InitializeObjectAttributes(&objectAttributes, &sourcePath, OBJ_CASE_INSENSITIVE, ObDosDevicesDirectory(), NULL);
+
+    status = NtOpenFile(
+            &sourceHandle,
+            FILE_GENERIC_READ,
+            &objectAttributes,
+            &ioStatusBlock,
+            FILE_SHARE_READ,
+            FILE_READ_ATTRIBUTES | FILE_SYNCHRONOUS_IO_NONALERT | FILE_OPEN_FOR_BACKUP_INTENT | FILE_NO_EA_KNOWLEDGE);
+    if (!NT_SUCCESS(status)) {
+        SetLastError(RtlNtStatusToDosError(status));
+        return FALSE;
+    }
+
+    // Fail overwriting the file if the target == the source. This is
+    // intentionally done after the check for existence of the source
+    // to set the most appropriate LastError value.
+    if (!strcmp(lpExistingFileName, lpNewFileName)) {
+      NtClose(sourceHandle);
+      SetLastError(ERROR_SHARING_VIOLATION);
+      return FALSE;
+    }
+
+    status = NtQueryInformationFile(
+            sourceHandle,
+            &ioStatusBlock,
+            &networkOpenInformation,
+            sizeof(networkOpenInformation),
+            FileNetworkOpenInformation);
+    if (!NT_SUCCESS(status)) {
+        DWORD lastError = RtlNtStatusToDosError(status);
+        SetLastError(lastError);
+        return FALSE;
+    }
+
+    RtlInitAnsiString(&targetPath, lpNewFileName);
+    InitializeObjectAttributes(&objectAttributes, &targetPath, OBJ_CASE_INSENSITIVE, ObDosDevicesDirectory(), NULL);
+
+    status = NtCreateFile(
+            &targetHandle,  // FileHandle
+            FILE_GENERIC_WRITE,   // DesiredAccess
+            &objectAttributes, // ObjectAttributes
+            &ioStatusBlock,   // IoStatusBlock
+            &networkOpenInformation.AllocationSize,  // AllocationSize
+            networkOpenInformation.FileAttributes,  // FileAttributes
+            FILE_SHARE_READ | FILE_SHARE_WRITE,  // ShareAccess
+            bFailIfExists ? FILE_CREATE : FILE_SUPERSEDE,  // CreateDisposition
+            FILE_SYNCHRONOUS_IO_NONALERT | FILE_OPEN_FOR_BACKUP_INTENT);  // CreateOptions
+    if (!NT_SUCCESS(status)) {
+        NtClose(sourceHandle);
+        DWORD lastError = RtlNtStatusToDosError(status);
+        SetLastError(lastError);
+        return FALSE;
+    }
+
+    readBuffer = MmAllocateContiguousMemory(readBufferSize);
+    if (!readBuffer) {
+        NtClose(sourceHandle);
+        NtClose(targetHandle);
+        SetLastError(ERROR_NOT_ENOUGH_MEMORY);
+        return FALSE;
+    }
+
+    while (TRUE) {
+        const BYTE *bufferPos = readBuffer;
+        if (!ReadFile(sourceHandle, readBuffer, readBufferSize, &bytesRead, NULL)) {
+            MmFreeContiguousMemory(readBuffer);
+            NtClose(sourceHandle);
+            NtClose(targetHandle);
+            return FALSE;
```

Comment:
Wouldn't these error-cases leave a broken file? Is that also how it works on Windows?

Reply:
They would. My reasoning was that an attempt to cleanup may spawn further failures and in the case that a file is copied into itself by bypassing the trivial check it might be dangerous to attempt to delete it.

Reply:
Good catch! I'd prefer to have the change in a follow-up PR, you can create an issue if you want for tracking it.

Reply:
Sounds good, filed #504